src/appfl/algorithm/scheduler/sync_scheduler.py

```diff
@@ -0,0 +1,46 @@
+from typing import Any, Union, Dict, OrderedDict
+from concurrent.futures import Future
+from omegaconf import DictConfig
+from appfl.algorithm.scheduler import BaseScheduler
+
+class SyncScheduler(BaseScheduler):
+    def __init__(
+        self, 
+        scheduler_configs: DictConfig, 
+        aggregator: Any,
+        logger: Any
+    ):
+        super().__init__(scheduler_configs, aggregator, logger)
+        self.local_models = {}
+        self.future = {}
+        self.num_clients = self.scheduler_configs.num_clients
+        self._num_global_epochs = 0
+
+    def schedule(self, client_id: Union[int, str], local_model: Union[Dict, OrderedDict], **kwargs) -> Future:
+        """
+        Schedule a synchronous global aggregation for the local model from a client.
+        The method will return a future object for the aggregated model, which will
+        be set after all clients have submitted their local models for the global aggregation.
+        :param client_id: the id of the client
+        :param local_model: the local model from a client
+        :param kwargs: additional keyword arguments for the scheduler
+        :return: the future object for the aggregated model
+        """
+        future = Future()
```

Comment:
Where do we set the number of threads? Does this assume that the server always runs on CPU?

Reply:
This does not assume the server has to run on CPU. The number of threads supported depends on the communicator used, for example, for grpc, the threads are specified here: https://github.com/APPFL/APPFL/blob/63810d2350dc51c1f3f64a031f3f1a915e7c13f8/src/appfl/comm/grpc/serve.py#L26

Reply:
Hi @kibaekkim, here are some conclusions I have regarding the usage of `concurrent.futures.Future`:

**TL;DR**: The usage of the Future object in all the schedulers does not restrict the server device to CPU-only, and there is no need to set a `max_workers` for that.

**Detailed Step-by-step Explanations**:

(1) What is the relationship between `concurrent.futures.Future` and `concurrent.futures.ThreadPoolExecutor`?

Answer: `ThreadPoolExecutor(max_workers=10)` provides an executor that you can submit as many concurrent functions to run as you want in an asynchronous way (note: at most 10 running concurrently, and if you submit more at the same time, some will be queued), and the running of these functions do not block the main program. As a consequence, the result of `executor.submit` is of type `Future`, and will be set to the actual results when the asynchronous function finishes running.

(2) If we directly create a `Future` object and set its value, does any concurrency involved?

Answer: No, for example, for the following code, everything runs serially.
```python
future = Future()
print(future.done()) # False - as no result is set for it
future.set_result(42)
print(future.done()) # True
print(future.result()) # 42
```

(3) How does the scheduler utilize the `concurrent.futures.Future` object?

Answer: Same as the above example, the scheduler module creates a `Future` object to indicate that the results (i.e., the global model) are not yet available - there are no concurrency/multi-thread involved in this process.

(4) Okay, so no concurrency involved, why we have `futures.ThreadPoolExecutor(max_workers=max_workers)` within the gRPC server?

Answer: gRPC uses this `ThreadPoolExecutor` to launch multiple threads to be able to handle requests from multiple clients at the same time. In this case, each thread gets a `Future` object after calling the sync scheduling function.

(5) Why the MPI server does not need multiple threads?

Answer: In our MPI communicator, instead of handling request for multiple clients at the same time, it receives client’s local model one by one from its receiving queue, passes to the scheduler, and gets a `Future` object. All those `Future` objects will be set after receiving the last model and performing the global aggregation. The set `Future` objects (i.e., new global model) will be send back to clients.

(6) So gRPC server uses multi thread while MPI server does not, does this affect the efficiency?

Answer: No for two reasons.

- All the scheduling logics have critical section that are protected by locks (e.g., [sync](https://github.com/APPFL/APPFL/blob/b3932ea32874c0d51ac44c12b4d3d3bec309079d/src/appfl/algorithm/scheduler/sync_scheduler.py#L33) and should be executed serially, so even with multiple threads, the scheduler runs serially.
- The scheduling can be described as IO-bound, i.e., its efficiency depends on when the last model will be sent to the server, so the concurrency will not have large increase to the overall efficiency.